README.rst
.. highlight:: sh

gramine-sgx-otk
***************

Introduction
============

Intel SGX requires RSA signature for each enclave that is to be executed,
irrespective of if the signature is actually needed by the attestation (that is,
if you attest against ``ENCLAVEHASH`` aka ``MRENCLAVE``, not against
``MRSIGNER``). Even if it's not really needed, CPU requires that the signature
is present and valid, and leakage of the RSA private key may compromise the data
protected by the enclave. Therefore the private key needs to be kept secure, in
spite of the fact that is serves no purpose for its holder.

This signing application generates random RSA keys and provably discards them
after a single operation. After the signing, there's no risk that the private
key leaks.

Quickstart
==========

On Debian 11 (``bullseye``):

::

    # add gramine repository, if you didn't already
    sudo curl -fsSLo /usr/share/keyrings/gramine-keyring.gpg \
        https://packages.gramineproject.io/gramine-keyring.gpg
    echo "deb [arch=amd64 signed-by=/usr/share/keyrings/gramine-keyring.gpg] https://packages.gramineproject.io/ bullseye main" \
    | sudo tee /etc/apt/sources.list.d/gramine.list

    sudo apt-get build-dep . -t bullseye-backports
    debuild

    sudo apt-get update
    sudo apt-get install ../gramine-sgx-otk_*.deb
    sudo gramine-sgx-otk init

    gramine-sgx-otk sign --inplace path/to/sigstruct.sig
    gramine-sgx-otk get-quote --sigstruct path/to/sigstruct.sig


debian/changelog
gramine-sgx-otk (0.1~UNRELEASED) UNRELEASED; urgency=medium

  *

 -- Wojtek Porczyk <woju@invisiblethingslab.com>  Wed,  1 Mar 2023 15:40:00 +0100


graminelibos_otk/store.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2023 Wojtek Porczyk <woju@invisiblethingslab.com>

import os
import pathlib

from . import structs

_XDG_CONFIG_HOME = pathlib.Path(
    os.getenv('XDG_CONFIG_HOME', os.path.expanduser('~/.config')))

QUOTE_STORE = _XDG_CONFIG_HOME / 'gramine' / 'otk-quotes'


class QuoteLookupError(LookupError):
    pass

class InvalidQuoteError(Exception):
    pass

class QuoteStore:
    def __init__(self, path=QUOTE_STORE):
        self.path = pathlib.Path(path)

    def get_quote_for_sigstruct(self, sigstruct):
        return self.get_quote_for_mrsigner(
            structs.Sigstruct(sigstruct).get_mrsigner())

    def get_quote_for_modulus(self, modulus):
        return self.get_quote_for_mrsigner(
            structs.get_mrsigner_for_modulus(modulus))

    def get_quote_for_mrsigner(self, mrsigner):
        try:
            with open(self.path) as file:
                for line in file:
                    quote = structs.Quote.fromhex(line)
                    if quote.report_data_mrsigner == mrsigner:
                        return quote
        except FileNotFoundError:
            pass # will raise below
        raise QuoteLookupError('no matching quote found')

    def save(self, modulus, quote):
        quote = structs.Quote(quote)
        expected_mrsigner = structs.get_mrsigner_for_modulus(modulus)
        found_mrsigner = quote.report_data_mrsigner
        if found_mrsigner != expected_mrsigner:
            raise InvalidQuoteError(
                f'MRSIGNER in the quote ({found_mrsigner.hex()}) does not '
                f'match the intended modulus ({expected_mrsigner.hex()})')

        self.path.parent.mkdir(parents=True, exist_ok=True)
        with open(self.path, 'a') as file:
            print(quote.hex(), file=file)


# vim: tw=80 ts=4 sts=4 sw=4 et


graminelibos_otk/signer.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2023 Wojtek Porczyk <woju@invisiblethingslab.com>

import functools
import pathlib
import shutil
import subprocess
import sys
import tempfile

import click

from . import (
    structs,
    store,
)

#: Path to default signer application directory.
DEFAULT_APPDIR = '/var/lib/gramine-sgx-otk'

#: Path to default template that will be rendered into signer application.
DEFAULT_TEMPLATE = '/usr/share/gramine-sgx-otk/gramine-sgx-otk.manifest.jinja'


def init(appdir=DEFAULT_APPDIR, template=DEFAULT_TEMPLATE, manifest_args=()):
    """
    Initialise the signing application.

    This function creates the directory for the new application (unlinking
    anything that was there previously), renders the manifest and signs it with
    a temporary key (though this temporary key can't be proved to have been
    discarded).

    Args:
        appdir (pathlib.Path): path to the application
        template (pathlib.Path): path to the template to be rendered for app
        manifest_args (iterable of strings): command line arguments to
            gramine-manifest
    """
    appdir = pathlib.Path(appdir)
    try:
        shutil.rmtree(appdir)
    except FileNotFoundError:
        pass
    appdir.mkdir(parents=True)

    manifest = appdir / 'gramine-sgx-otk.manifest'

    subprocess.run(['gramine-manifest',
        *manifest_args,
        template,
        manifest
    ], check=True)

    return update_measurement(appdir)


def update_measurement(appdir):
    """
    Update measurement of the signing application.

    Given the application directory, it re-measures the files and re-signs the
    manifest using ``gramine-sgx`` using a temporary key (though this temporary
    key can't be proved to have been discarded). The manifest is not
    re-rendered.

    Args:
        appdir (pathlib.Path): path to the application
    """
    appdir = pathlib.Path(appdir)
    manifest = appdir / 'gramine-sgx-otk.manifest'

    with tempfile.NamedTemporaryFile() as tmpkey:
        subprocess.run(['gramine-sgx-gen-private-key', '-f', tmpkey.name])
        subprocess.run(['gramine-sgx-sign',
            '--key', tmpkey.name,
            '--manifest', manifest,
            '--output', manifest.with_suffix('.manifest.sgx')
        ])

    # if we're on EPID, we need to unlink token, because the automation in
    # gramine-sgx launcher does not check validity of the token
    (appdir / 'gramine-sgx-otk.token').unlink(missing_ok=True)

    # TODO: record mrenclave, gramine version, Cargo.lock, openssl version


def show_quote_cb(modulus, quote):
    """
    Callback that shows modulus and quote during signing.

    This callback is suitable for use with *quote_cb* argument to
    :func:`sign_with_otk_cb`. *modulus* and *quote* are formatted
    and printed to standard error.
    """
    print('modulus:', file=sys.stderr)
    for line in structs.hexdump(modulus):
        print(f'    {line}', file=sys.stderr)
    print('mrsigner:', file=sys.stderr)
    for line in structs.hexdump(structs.get_mrsigner_for_modulus(modulus)):
        print(f'    {line}', file=sys.stderr)
    print('quote:', file=sys.stderr)
    for line in structs.hexdump(quote):
        print(f'    {line}', file=sys.stderr)


@click.command(add_help_option=False)
@click.help_option('--help-otk')
@click.option('--otk-appdir', 'appdir',
    type=click.Path(dir_okay=True, file_okay=False),
    default=DEFAULT_APPDIR,
    help='Path to signing application, where gramine manifest is stored.')
@click.option('--otk-check-isvsvn/--otk-no-check-isvsvn', 'check_isvsvn',
    default=False,
    help='Check if ISVSVN is at maximum value (0xFFFF).')
@click.option('--otk-show-quote/--otk-no-show-quote',
    default=False,
    help='Show the modulus and the quote to stderr (for debugging).')
def sign_with_otk(otk_show_quote, **kwds):
    return functools.partial(sign_with_otk_cb,
        quote_cb=(show_quote_cb if otk_show_quote else None),
        **kwds)


def sign_with_otk_cb(data, *,
        appdir=DEFAULT_APPDIR,
        quote_cb=None,
        check_isvsvn=False,
        _do_not_store_quote=False):
    """
    Signing function for signing SGX enclaves.

    *data* is expected to be signing data extracted from SIGSTRUCT.

    If *quote_cb* is not ``None``, it needs to be 2-argument function that will
    be called with *modulus* and *quote* (instances of ``bytes``, *modulus* is
    big endian).
    """
    if check_isvsvn:
        isvsvn = int.from_bytes(data[254:256], 'little')
        if isvsvn != structs.MAX_ISVSVN:
            raise structs.InvalidSigstruct(
                f'expected ISVSVN {structs.MAX_ISVSVN}, found {isvsvn:#06x}')

    try:
        proc = subprocess.run(['gramine-sgx', 'gramine-sgx-otk'],
            cwd=appdir, input=data, capture_output=True, check=True)

    except subprocess.CalledProcessError as e:
        sys.stderr.buffer.write(e.stderr)
        print('failed to sign', file=sys.stderr)
        sys.exit(1)

    sys.stderr.buffer.write(proc.stderr)
    sys.stderr.buffer.flush()
    modulus =   proc.stdout[0  :384]
    signature = proc.stdout[384:768]
    quote =     proc.stdout[768:]

    if quote_cb is not None:
        quote_cb(modulus, quote)

    if not _do_not_store_quote:
        store.QuoteStore().save(modulus, quote)

    return (
        3,
        int.from_bytes(modulus, byteorder='big'),
        int.from_bytes(signature, byteorder='big'),
    )


def sign(sigstruct, **kwds):
    assert len(sigstruct) == structs.SIGSTRUCT_SZ
    sigstruct = structs.Sigstruct(sigstruct)
    signing_data = sigstruct.get_signing_data()
    exponent, modulus, signature = sign_with_otk_cb(signing_data, **kwds)
    assert exponent == 3

    (sigstruct.exponent, sigstruct.modulus, sigstruct.signature
        ) = exponent, modulus, signature

    # https://eprint.iacr.org/2016/086.pdf, 6.5.2
    sigstruct.q1, w = divmod(signature ** 2, modulus)
    sigstruct.q2 = w * signature // modulus

    return sigstruct


# vim: tw=80 ts=4 sts=4 sw=4 et


graminelibos_otk/structs.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2023 Wojtek Porczyk <woju@invisiblethingslab.com>

import hashlib
import itertools
import struct

SIGSTRUCT_SZ = 1808
SIGSTRUCT_HEADER  = bytes.fromhex('06000000E10000000000010000000000')
SIGSTRUCT_HEADER2 = bytes.fromhex('01010000600000006000000001000000')

MAX_ISVSVN = 0xffff


def hexdump(data, width=64):
    for _, line in itertools.groupby(enumerate(data.hex()),
            key=(lambda e: e[0] // width)):
        yield ''.join(c for _, c in line)


def get_mrsigner_for_modulus(modulus):
    return get_mrsigner_for_modulus_le(bytes(reversed(modulus)))

def get_mrsigner_for_modulus_le(modulus):
    return hashlib.sha256(modulus).digest()


class _field_ro:
    class Value(bytes):
        def __int__(self):
            return int.from_bytes(self, byteorder='little')

    def __init__(self, offset, length):
        self.offset = offset
        self.length = length
        self.struct = struct.Struct(f'{length}s')

    def __get__(self, instance, owner=None):
        if instance is None:
            return self
        return self.Value(self.struct.unpack_from(instance, self.offset)[0])

    def __set__(self, instance, value):
        raise AttributeError('can\'t set attribute')

class _field(_field_ro):
    def __set__(self, instance, value):
        if isinstance(value, int):
            value = value.to_bytes(length=self.length, byteorder='little')
        self.struct.pack_into(instance, self.offset, value)


class InvalidSigstruct(Exception):
    pass

class Sigstruct(bytearray):
    # SDM vol. 3D part 4, 38.13, table 38-19
    # https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3d-part-4-manual.pdf
    header =    _field(   0,  16)
    header2 =   _field(  24,  16)
    modulus =   _field( 128, 384)
    exponent =  _field( 512,   4)
    signature = _field( 516, 384)
    mrenclave = _field( 960,  32)
    isvprodid = _field(1024,   2)
    isvsvn =    _field(1026,   2)
    q1 =        _field(1040, 384)
    q2 =        _field(1424, 384)

    def __init__(self, *args, **kwds):
        super().__init__(*args, **kwds)

        if len(self) != SIGSTRUCT_SZ:
            raise InvalidSigstruct(
                f'wrong length: expected {SIGSTRUCT_SZ}, got {len(self)}')

        # Prevent bytearray from resizing by holding memoryview of ourselves.
        # As long as memoryview lives, all operations that would change the
        # length, like bytearray.append(), will instead raise BufferError.
        # see also:
        # https://docs.python.org/3/library/stdtypes.html#memoryview.release
        self._m = memoryview(self)

        if self.header != SIGSTRUCT_HEADER:
            raise InvalidSigstruct('wrong HEADER')
        if self.header2 != SIGSTRUCT_HEADER2:
            raise InvalidSigstruct('wrong HEADER2')

    @classmethod
    def from_file(cls, file):
        return cls(file.read(SIGSTRUCT_SZ + 1))

    def get_signing_data(self):
        return self[0:128] + self[900:1028]

    def get_mrsigner(self):
        return get_mrsigner_for_modulus_le(self.modulus)


class Quote(bytes):
    report_data_mrsigner =  _field_ro(368, 32)
    mrenclave =             _field_ro(112, 32)


graminelibos_otk/__main__.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2023 Wojtek Porczyk <woju@invisiblethingslab.com>

import pathlib

import click

from . import (
    signer,
    structs,
    store,
)


@click.group()
def cli():
    pass


@cli.command()
@click.option('--appdir', metavar='PATH',
    type=click.Path(dir_okay=True, file_okay=False),
    default=signer.DEFAULT_APPDIR,
    help='Path to signing application, where gramine manifest is stored.')
@click.option('--check-isvsvn/--no-check-isvsvn',
    default=False,
    help='Check if ISVSVN is at maximum value (0xFFFF).')
@click.option('--show-quote/--no-show-quote',
    default=False,
    help='Show the modulus and the quote to stderr (for debugging).')
@click.option('--output', metavar='PATH',
    type=click.File('wb'),
    help='Where to write the signed SIGSTRUCT.')
@click.option('--inplace/--no-inplace',
    default=False,
    help='Overwrite the SIGSTRUCT in place.')
@click.argument('file', metavar='PATH',
    type=click.File('r+b'))
def sign(appdir, check_isvsvn, show_quote, output, inplace, file):
    """
    Sign an existing SIGSTRUCT.

    The SIGSTRUCT should be created elsewhere, possibly by gramine-sgx-sign.
    """

    sigstruct = signer.sign(structs.Sigstruct.from_file(file),
        appdir=pathlib.Path(appdir),
        quote_cb=(signer.show_quote_cb if show_quote else None),
        check_isvsvn=check_isvsvn)

    if output:
        output.write(sigstruct)

    if inplace:
        file.seek(0)
        file.truncate()
        file.write(sigstruct)


def cb_unhexlify(_ctx, _param, value):
    return value and bytes.fromhex(value)


@cli.command()
@click.pass_context
@click.option('--sigstruct', 'sigstruct_file', metavar='PATH',
    type=click.File('rb'),
    help='Read SIGSTRUCT from file, extract modulus and calculate MRSIGNER.')
@click.option('--modulus-be', metavar='HEX',
    callback=cb_unhexlify,
    help='Calculate MRSIGNER based on given modulus (big endian).')
@click.option('--modulus-le', metavar='HEX',
    callback=cb_unhexlify,
    help='Calculate MRSIGNER based on given modulus'
        ' (little endian, like in the SIGSTRUCT.')
@click.option('--mrsigner', metavar='HEX',
    callback=cb_unhexlify,
    help='Query for given MRSIGNER.')
@click.argument('quote_file', metavar='PATH',
    type=click.File('wb'), default='-')
def get_quote(ctx, quote_file,
        sigstruct_file, modulus_be, modulus_le, mrsigner):
    """
    Get quote for a signing key.

    The signing key can be specified using a variety of methods:
    """
    if sum(int(opt is not None) for opt in (
        sigstruct_file,
        modulus_be,
        modulus_le,
        mrsigner,
    )) != 1:
        ctx.fail('specify exactly one of:'
            ' --sigstruct'
            ' --modulus-be'
            ' --modulus-le'
            ' --mrsigner'
        )

    quote_store = store.QuoteStore()

    try:
        if sigstruct_file is not None:
            sigstruct = structs.Sigstruct.from_file(sigstruct_file)
            quote = quote_store.get_quote_for_sigstruct(sigstruct)
        elif modulus_be is not None:
            quote = quote_store.get_quote_for_modulus(modulus_be)
        elif modulus_le is not None:
            quote = quote_store.get_quote_for_modulus(bytes(reversed(modulus_be)))
        elif mrsigner is not None:
            quote = quote_store.get_quote_for_mrsigner(mrsigner)

    except store.QuoteLookupError as e:
        click.echo(str(e), err=True)
        ctx.exit(1)

    except Exception as e:
        ctx.fail(str(e))

    quote_file.write(quote)


@cli.command()
@click.option('--appdir', metavar='PATH',
    type=click.Path(dir_okay=True, file_okay=False),
    default=signer.DEFAULT_APPDIR,
    help='Path to signing application, where gramine manifest is stored.')
@click.option('--template', metavar='PATH',
    type=click.Path(exists=True, dir_okay=False),
    default=signer.DEFAULT_TEMPLATE,
    help='Path to manifest template.')
@click.argument('manifest-args', metavar='ARGS', nargs=-1)
def init(appdir, template, manifest_args):
    """
    Initialise the signing application.

    ARGS are passed directly to gramine-manifest, and this is where you
    configure your application (debug mode, EPID credentials, ...).
    """
    signer.init(appdir, template, manifest_args)


@cli.command()
@click.option('--appdir', metavar='PATH',
    type=click.Path(dir_okay=True, file_okay=False),
    default=signer.DEFAULT_APPDIR,
    help='Path to signing application, where gramine manifest is stored.')
def update_measurement(appdir):
    """
    Update measurement of the signing application.

    This is needed in case some dependencies od the application (like openssl)
    were changed.
    """
    signer.update_measurement(appdir)


if __name__ == '__main__':
    cli()

# vim: tw=80 ts=4 sts=4 sw=4 et
